### ISSUE TITLE
Allow Volume.export() a coroutine

### ISSUE THREAD

**Issue by marmarek** [QubesOS/qubes-issues#5935]
**The problem you're addressing (if any)**
If `Volume.export()` function needs to make some changes (mount something, do a qrexec call or else), it needs to be a coroutine to not lock the whole qubesd process.

**Describe the solution you'd like**
Allow `Volume.export()` to be a coroutine, pretty much like every other function in the storage API. This requires changes in places that do use `Volume.export()` - especially backup code.
Since `Volume.export()` could make some changes, it would be useful also to add `Volume.export_end()` (similar as it's done for `Volume.import_data` already) to undo those changes.

**Where is the value to a user, and who might that user be?**
Useful at least for https://github.com/QubesOS/qubes-core-admin/pull/354

**Comment by marmarek** [comment]
@rustybird this requires making `ReflinkVolume.import_volume` a real coroutine, not a synchronous function run in a separate thread as a whole (otherwise you can't write `await/yield from src_volume.export()`). This also makes `threading.Lock` usage inconvenient at least.
I've opened a draft PR with other changes related to this. I'll change reflink driver to handle it, just FYI. If you want to handle this somehow else, let me know.

**Comment by rustybird** [comment]
> This also makes `threading.Lock` usage inconvenient at least.

Yeah... the best I can come up with is something like:

```py
    @asyncio.coroutine
    def import_volume(self, src_volume):
        if self.save_on_stop:
            if not self._lock.acquire(blocking=False):
                yield from _coroutinized(self._lock.acquire)()
            try:
                try:
                    success = False
                    src_path = yield from qubes.utils.coro_maybe(
                        src_volume.export())
                    try:
                        yield from _coroutinized(_copy_file)(src_path,
                                                             self._path_import)
                    finally:
                        yield from qubes.utils.coro_maybe(
                            src_volume.export_end(src_path))
                    success = True
                finally:
                    yield from _coroutinized(self._import_data_end)(success)
            finally:
                self._lock.release()
        return self
```

**Comment by marmarek** [comment]
I would use a lock from asyncio, which should fit nicely here. There is one little issue - `_get_size` which can't be a coroutine (a property getter) but it uses a lock. Why exactly is that? Especially when very similar `usage` getter don't have that lock. Can it be dropped in `_get_size` too?

**Comment by marmarek** [comment]
It's also tempting to start using `async def` syntax instead of `@asyncio.coroutine` decorator, but it means backports to R4.0 will be harder.

**Comment by rustybird** [comment]
> There is one little issue - `_get_size` which can't be a coroutine (a property getter) but it uses a lock. Why exactly is that? Especially when very similar `usage` getter don't have that lock.

The `size` getter also updates the nominal `self._size`. If it's running in parallel with `resize` it could clobber that variable, so e.g. a volatile volume would go back to the old size on the next `start`.

**Comment by marmarek** [comment]
Hmm, that's weird for a property getter to change the state... I see it doesn't prefer the cached value, so it isn't really performance optimization. Maybe setting `self._size` can be dropped there? Return a file size (if file present) and fallback to `self._size` if not - but do not update it?

**Comment by rustybird** [comment]
The getter updates `_size` out of a [general paranoia](https://github.com/QubesOS/qubes-core-admin/commit/2b4b45e) of ever going back to a stale nominal size. But the only scenario that comes to mind where it makes any difference is this: The user (1.) manually resizes a volatile image file, then (2.) looks at its `qvm-volume info` output - "aha, Qubes has picked up the new size", and finally (3.) restarts the volatile volume, which would recreate it with the old size if the getter didn't update in step 2. => Meh, somewhat unexpected but probably not worth catering for if it results in ugly locking code?

I'll submit a PR that changes the getter and also moves the lvm `@locked` decorator to a `Volume` base class `@staticmethod` so reflink can use it too, if that sounds good to you.

**Comment by marmarek** [comment]
In that case, perhaps better update `_size` in `ReflinkVolume.stop()`?

> I'll submit a PR that changes the getter and also moves the lvm `@locked` decorator to a `Volume` base class `@staticmethod` so reflink can use it too, if that sounds good to you.

Yes, that sounds good.

**Comment by rustybird** [comment]
> In that case, perhaps better update `_size` in `ReflinkVolume.stop()`?

Thanks for the hint! Included in QubesOS/qubes-core-admin#357.
